### PR TITLE
obj-tiler : Adding support for meshes with materials but no texture. 

### PR DESCRIPTION
**Pull request by charly-perspectives**
Hey, 
I miss data to test it on .obj with textures.

### CODE REVIEW

File: py3dtilers/ObjTiler/obj.py [hunk 1]
```diff
@@ -110,22 +114,31 @@ def retrieve_objs(files, with_texture=False):
...
-
         :return: a list of Obj.
         """
         objects = list()
 
         for obj_file in files:
             print("Reading " + str(obj_file))
-            geom = pywavefront.Wavefront(obj_file, collect_faces=True)
+            geom = pywavefront.Wavefront(obj_file, collect_faces=True, create_materials=True)
+            mesh = geom.mesh_list[0]
```
Comment (LorenzoMarnat) [inline py3dtilers/ObjTiler/obj.py:124]: > mesh = geom.mesh_list[0]

We need to iterate on all the meshes in the obj file, so we can delete this line (see comment below)

File: py3dtilers/ObjTiler/obj.py [hunk 2]
```diff
@@ -110,22 +114,31 @@ def retrieve_objs(files, with_texture=False):
...
             if len(geom.vertices) == 0:
                 continue
-            for mesh in geom.mesh_list:
-                # Get id from its name
+            gltfMaterials = []
+            mesh_index = 1
+
+            for mesh in mesh.materials:
```
Comment (LorenzoMarnat) [inline py3dtilers/ObjTiler/obj.py:130]: We need to iterate on all meshes then all materials, so we need a double for loop:  

```python
            for mesh in geom.mesh_list:
                for mesh_mat in mesh.materials:
                    # get id from its name
                    id = mesh.name + str(mesh_index)
                    obj = Obj(id)
                    obj.set_material_index(mesh_index)
                    mesh_index += 1
                    if obj.parse_geom(mesh_mat, with_texture):
                        objects.append(obj)
                    material = GlTFMaterial(rgb=[mesh_mat.diffuse[0], mesh_mat.diffuse[1], mesh_mat.diffuse[2]], alpha=1. - mesh_mat.diffuse[3], metallicFactor=0.)
                    gltfMaterials.append(material)
```

File: py3dtilers/ObjTiler/obj.py [hunk 3]
```diff
@@ -110,22 +114,31 @@ def retrieve_objs(files, with_texture=False):
...
+                # get id from its name
                 id = mesh.name
                 obj = Obj(id)
-                # Create geometry as expected from GLTF from an obj file
-                if obj.parse_geom(mesh, with_texture):
+                obj.set_material_index(mesh_index)
+                mesh_index += 1
+                if obj.parse_geom(mesh, with_texture):
```
Comment (LorenzoMarnat) [inline py3dtilers/ObjTiler/obj.py:136]: Delete trailing whitespaces after colon

**Review by LorenzoMarnat — CHANGES_REQUESTED** [review]
Hello, thank you for your contribution !

There are some changes to do but it seems to work fine

If you want to test the tiler with free obj models, you can find some online (for example [this one](https://free3d.com/3d-model/e-45-aircraft-71823.html))

**Comment by charly-perspectives** [comment]
Hey!
Thanks for the feedback. I think I made all the changes.
About the tests on textured .obj: it's just that I use itowns to visualize tilesets, which requires geolocalized models.

**Comment by LorenzoMarnat** [comment]
> About the tests on textured .obj: it's just that I use itowns to visualize tilesets, which requires geolocalized models.

If you want to geolocalized your obj model, you can use the `--offset x y z` flag to put your model at an arbitrary position. For example, to put my test obj model in Lyon city in EPSG:3946 projection (to be able to visualize it with itowns), I use:

```bash
obj-tiler -i E:\Data\Obj\E-45-Aircraft\ -o E:\MyRepo --offset 1843397 5173891 300 --scale 20 --with_texture
```

File: py3dtilers/ObjTiler/obj.py [hunk 4]
```diff
@@ -110,22 +114,31 @@ def retrieve_objs(files, with_texture=False):
...
-        return Objs(objects)
+                for mesh_mat in mesh.materials:
+                    # get id from its name
+                    id = mesh_mat.name
+                    obj = Obj(id)
+                    obj.set_material_index(mesh_index)
+                    mesh_index += 1
+                    if obj.parse_geom(mesh_mat, with_texture):
+                        objects.append(obj)
+                    material = GlTFMaterial(rgb=[mesh.diffuse[0], mesh.diffuse[1], mesh.diffuse[2]], alpha=1. - mesh.diffuse[3], metallicFactor=0.)
```
Comment (LorenzoMarnat) [inline py3dtilers/ObjTiler/obj.py:138]: One last change: in line 138, all occurences of `mesh` should be `mesh_mat`

Comment (charly-perspectives) [reply py3dtilers/ObjTiler/obj.py]: my bad. it should be fine now.

**Pull request merged**